packages/graphql/lib/src/link/link.dart

```diff
@@ -12,26 +12,42 @@ typedef RequestHandler = Stream<FetchResult> Function(
   NextLink forward,
 ]);
 
-Link _concat(
-  Link first,
-  Link second,
-) {
-  return Link(request: (
-    Operation operation, [
-    NextLink forward,
-  ]) {
-    return first.request(operation, (Operation op) {
-      return second.request(op, forward);
-    });
-  });
+abstract class Link {
+  static Link from(List<Link> links) => links.reduce(
+        (first, next) => first.concat(next),
+      );
+
+  static Link fromHandler(RequestHandler handler) =>
+      _RequestHandlerLink(handler);
+
+  Stream<FetchResult> request(Operation operation, [NextLink forward]);
+
+  Link concat(Link next) => _ConcatLink(this, next);
+}
+
+class _RequestHandlerLink extends Link {
```

Comment:
The spirit of this is good, but we can do this in both a completely backwards-compatible and more Apollo compliant way:
* `_RequestHandlerLink` was the publicly exposed `Link`, with an api like
```dart
Link({RequestHandler request}): this._handler = request
```
* dump all the `abstract class Link` static methods into `_RequestHandlerLink`
* add some safety in `request` that's something like
```dart
assert(_handler != null, "Custom Link $runtimeType has no RequestHandler")
```